src/networking.c

```diff
@@ -1347,6 +1352,105 @@ void unlinkClient(client *c) {
+    }
+
+    serverAssert(!(c->flags &(CLIENT_SLAVE|CLIENT_MASTER)));
+
+    if (c->flags & CLIENT_TRACKING) disableTracking(c);
+    selectDb(c,0);
+    c->resp = 2;
+
+    clientSetDefaultAuth(c);
+    moduleNotifyUserChanged(c);
+    discardTransaction(c);
+
+    pubsubUnsubscribeAllChannels(c,0);
+    pubsubUnsubscribeAllPatterns(c,0);
+
+    if (c->name) {
+        decrRefCount(c->name);
+        c->name = NULL;
+    }
+
+    /* Selectively clear state flags not covered above */
+    c->flags &= ~(CLIENT_ASKING|CLIENT_READONLY|CLIENT_PUBSUB|
+                  CLIENT_REPLY_OFF|CLIENT_REPLY_SKIP_NEXT);
+}
+
+/* Clear the client state fully and initialize it as a new client object. This
+ * function is useful when you want to reuse the client object to avoid
+ * costly freeClient()/createClient() calls. */
+void clearClient(client *c) {
+    serverAssert(!(c->flags & CLIENT_PROTECTED));
+
+    /* For connected clients, call the disconnection event of modules hooks. */
+    if (c->conn) {
+        moduleFireServerEvent(REDISMODULE_EVENT_CLIENT_CHANGE,
+                              REDISMODULE_SUBEVENT_CLIENT_CHANGE_DISCONNECTED,
+                              c);
+    }
+
+    clearClientConnectionState(c);
+
+    /* Deallocate structures used to block on blocking ops. */
+    if (c->flags & CLIENT_BLOCKED) unblockClient(c);
+    dictEmpty(c->bpop.keys, NULL);
+
+    sdsclear(c->querybuf);
+    /* Resize querybuf if it is too big. If this client ends up in a cache
+     * after being cleared, it may go undetected with a large query buffer,
+     * possibly wasting memory. */
+    if (sdsavail(c->querybuf) > PROTO_RESIZE_THRESHOLD) {
+        c->querybuf = sdsResize(c->querybuf, PROTO_RESIZE_THRESHOLD);
+    }
+
+    sdsclear(c->pending_querybuf);
+    listEmpty(c->watched_keys);
+    dictEmpty(c->pubsub_channels, NULL);
+    listEmpty(c->pubsub_patterns);
+    listEmpty(c->reply);
+    freeReplicaReferencedReplBuffer(c);
+    freeClientArgv(c);
+    freeClientOriginalArgv(c);
+
+    /* Unlink the client: this will close the socket, remove the I/O
+     * handlers, and remove references of the client from different
+     * places where active clients may be referenced. */
+    unlinkClient(c);
+
+    /* Remove the contribution that this client gave to our
+     * incrementally computed memory usage. */
+    server.stat_clients_type_memory[c->last_memory_type] -=
+        c->last_memory_usage;
+    /* Remove client from memory usage buckets */
+    if (c->mem_usage_bucket) {
+        c->mem_usage_bucket->mem_usage_sum -= c->last_memory_usage;
+        listDelNode(c->mem_usage_bucket->clients, c->mem_usage_bucket_node);
+    }
+
+    /* Release other dynamically allocated client structure fields */
+    if (c->name) decrRefCount(c->name);
+    sdsfree(c->peerid);
+    sdsfree(c->sockname);
+    sdsfree(c->slave_addr);
+
+    /* Reinitialize all variables to the default values */
+    initClientVariables(c);
```

Comment:
this method is quite heavy.
ideally we could put all these fields together and reset them with a memset.
i can't think of a good way to do that though.

Reply:
I'll take a look. Probably, compiler is doing the same as most variables are initialized to zero. I expect it use vectorized instructions here.